src/data/polynomial/module.lean

```diff
@@ -0,0 +1,188 @@
+In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R[X]`-module `M[X]`.
+
+This is defined as an type alias `polynomial_module R M := ℕ →₀ M`, since there might be different
+module structures on `ℕ →₀ M` of interest. See the docstring of `polynomial_module` for details.
+
+-/
+
+universes u v
+
+open polynomial
+open_locale polynomial big_operators
+
+variables (R M : Type*) [comm_ring R] [add_comm_group M] [module R M] (I : ideal R)
+
+include R
+
+/--
+The `R[X]`-module `M[X]` for an `R`-module `M`.
+This is isomorphic (as an `R`-module) to `polynomial M` when `M` is a ring.
+
+We require all the module instances `module S (polynomial_module R M)` to factor through `R` except
+`module R[X] (polynomial_module R M)`.
+In this constraint, we have the following instances for example :
+- `R` acts on `polynomial_module R R[X]`
+- `R[X]` acts on `polynomial_module R R[X]` as `R[Y]` acting on `R[X][Y]`
+- `R` acts on `polynomial_module R[X] R[X]`
+- `R[X]` acts on `polynomial_module R[X] R[X]` as `R[X]` acting on `R[X][Y]`
+- `R[X][X]` acts on `polynomial_module R[X] R[X]` as `R[X][Y]` acting on itself
+
+This is also the reason why `R` is included in the alias, or else there will be two different
+instances of `module R[X] (polynomial_module R[X])`.
+
+See https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2315065.20polynomial.20modules
+for the full discussion.
+-/
+@[derive add_comm_group, derive inhabited, nolint unused_arguments]
+def polynomial_module := ℕ →₀ M
+
+omit R
+
+variables {M}
+variables {S : Type*} [comm_semiring S] [algebra S R] [module S M] [is_scalar_tower S R M]
+
+namespace polynomial_module
+
+/-- This is required to have the `is_scalar_tower S R M` instance to avoid diamonds. -/
+@[nolint unused_arguments]
+noncomputable
+instance : module S (polynomial_module R M) :=
+finsupp.module ℕ M
+
+instance : has_coe_to_fun (polynomial_module R M) (λ _, ℕ → M) :=
+finsupp.has_coe_to_fun
+
+/-- The monomial `m * x ^ i`. This is defeq to `finsupp.single_add_hom`, and is redefined here
+so that it has the desired type signature.  -/
+noncomputable
+def single (i : ℕ) : M →+ polynomial_module R M :=
+finsupp.single_add_hom i
+
+lemma single_apply (i : ℕ) (m : M) (n : ℕ) : single R i m n = ite (i = n) m 0 :=
+finsupp.single_apply
+
+/-- `polynomial_module.single` as a linear map. -/
+noncomputable
+def lsingle (i : ℕ) : M →ₗ[R] polynomial_module R M :=
+finsupp.lsingle i
+
+lemma lsingle_apply (i : ℕ) (m : M) (n : ℕ) : lsingle R i m n = ite (i = n) m 0 :=
+finsupp.single_apply
+
+variable {R}
+
+lemma induction_linear {P : polynomial_module R M → Prop} (f : polynomial_module R M)
+  (h0 : P 0) (hadd : ∀ f g, P f → P g → P (f + g)) (hsingle : ∀ a b, P (single R a b)) : P f :=
+finsupp.induction_linear f h0 hadd hsingle
+
+@[semireducible] noncomputable
+instance polynomial_module : module R[X] (polynomial_module R M) :=
+module_polynomial_of_endo ({ to_fun := λ (f : polynomial_module R M),
```

Comment:
It seems you can use [finsupp.lmap_domain](https://leanprover-community.github.io/mathlib_docs/linear_algebra/finsupp.html#finsupp.lmap_domain) here?

Can you show `M[X]` `is_tensor_product` of `R[X]` and `M` over `R`? Or better, it's isomorphic to R[X] ⊗ₜ[R] M as R[X]-modules (tensor_product.left_module)? Or that it's an extension of scalars (#15673 doesn't introduce such a predicate though)?

Reply:
I think [`is_base_change`](https://leanprover-community.github.io/mathlib_docs/ring_theory/is_tensor_product.html#is_base_change) is what you are looking for?
And it is indeed on my todo list - though it needs a bit more api around `is_base_change`, so it probably won't be in this PR.

Reply:
`finsupp.lmap_domain` works like a charm! Thanks!